wasm.cc

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (C) 2019-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "wasm.hh"
+#include "concrete_types.hh"
+#include "utils/utf8.hh"
+#include "utils/ascii.hh"
+#include "utils/date.h"
+#include "db/config.hh"
+#include <seastar/core/coroutine.hh>
+#include "seastarx.hh"
+
+static logging::logger wasm_logger("wasm");
+
+namespace wasm {
+
+  // Fuel defines more or less how many bytecode instructions
+  // can be performed at once
+  static constexpr uint64_t initial_fuel_amount = 1024*1024;
+
+static wasmtime::Config make_config() {
+    wasmtime::Config config;
+    config.consume_fuel(true);
+    return config;
+}
+
+//FIXME: Send sincere handwritten apologies to @xemul for introducing a global
+//       or drop the global in favor of a sharded service
+static wasmtime::Engine& wasm_engine() {
+    static thread_local wasmtime::Engine engine = wasmtime::Engine(make_config());
+    return engine;
```

Comment:
thread_local is not that bad (also not very good, but at least doesn't stand in the way of cluster-in-a-box).

Reply:
Still, better to have a sharded service so it can be configured with non-default config, if we ever need it.

Reply:
Yup, I decided to leave it like that and quote the thread_local usage from our schema_registry.cc as my main line of defense. This engine is such a thin layer that creating a sharded service out of it and propagating it all over the source code seems a little complicated without any immediate gain - especially that we compile in two separate layers - once when the function is created, and then when it arrives to another node during schema migration.

There are some configuration options which could be useful for exposure - e.g. max wasm stack size, but on the other hand they're fairly low level and we should probably come up with sensible defaults instead.

In this iteration I decided to accept this little debt, since if anybody needs custom configuration later, I'll be the one who implements the proper sharded service anyway, so at least I only indebt myself.